server/sonar-server/src/main/java/org/sonar/server/computation/filemove/FileMoveDetectionStep.java

```diff
@@ -0,0 +1,349 @@
+      // SourceHashRepository
+      Component component = reportFilesByKey.get(fileKey);
+      SourceLinesHashesComputer linesHashesComputer = new SourceLinesHashesComputer();
+      SourceHashComputer sourceHashComputer = new SourceHashComputer();
+      try (CloseableIterator<String> lineIterator = sourceLinesRepository.readLines(component)) {
+        while (lineIterator.hasNext()) {
+          String line = lineIterator.next();
+          linesHashesComputer.addLine(line);
+          sourceHashComputer.addLine(line, lineIterator.hasNext());
+        }
+      }
+      builder.put(fileKey, new File(component.getReportAttributes().getPath(), sourceHashComputer.getHash(), linesHashesComputer.getLineHashes()));
+    }
+    return builder.build();
+  }
+
+  private ScoreMatrix computeScoreMatrix(Map<String, DbComponent> dtosByKey, Set<String> dbFileKeys, Map<String, File> reportFileSourcesByKey) {
+    int[][] scoreMatrix = new int[dbFileKeys.size()][reportFileSourcesByKey.size()];
+    int maxScore = 0;
+
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      int dbFileIndex = 0;
+      for (String removedFileKey : dbFileKeys) {
+        File fileInDb = getFile(dbSession, dtosByKey.get(removedFileKey));
+        if (fileInDb == null) {
+          continue;
+        }
+
+        int reportFileIndex = 0;
+        for (Map.Entry<String, File> reportFileSourceAndKey : reportFileSourcesByKey.entrySet()) {
+          File unmatchedFile = reportFileSourceAndKey.getValue();
+          int score = fileSimilarity.score(fileInDb, unmatchedFile);
+          scoreMatrix[dbFileIndex][reportFileIndex] = score;
+          if (score > maxScore) {
+            maxScore = score;
+          }
+          reportFileIndex++;
+        }
+        dbFileIndex++;
+      }
+    }
+
+    return new ScoreMatrix(dbFileKeys, reportFileSourcesByKey, scoreMatrix, maxScore);
+  }
+
+  @CheckForNull
+  private File getFile(DbSession dbSession, DbComponent dbComponent) {
+    FileSourceDto fileSourceDto = dbClient.fileSourceDao().selectSourceByFileUuid(dbSession, dbComponent.getUuid());
+    if (fileSourceDto == null) {
+      return null;
+    }
+    return new File(dbComponent.getPath(), fileSourceDto.getSrcHash(), LINES_HASHES_SPLITTER.splitToList(fileSourceDto.getLineHashes()));
+  }
+
+  private static void printIfDebug(ScoreMatrix scoreMatrix) {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("ScoreMatrix:\n" + scoreMatrix.toCsv(';'));
+    }
+  }
+
+  private static ElectedMatches electMatches(Set<String> dbFileKeys, Map<String, File> reportFileSourcesByKey, MatchesByScore matchesByScore) {
+    ElectedMatches electedMatches = new ElectedMatches(matchesByScore, dbFileKeys, reportFileSourcesByKey);
+    Multimap<String, Match> matchesPerFileForScore = ArrayListMultimap.create();
+    for (List<Match> matches : matchesByScore) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the total number of break and continue statements in this loop to use at most one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS135)